src/EFCore.Analyzers/InterpolatedStringUsageInRawQueriesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,229 @@
+        var targetMethod = invocation.TargetMethod;
+
+        var report = targetMethod.Name switch
+        {
+            "FromSqlRaw" => AnalyzeFromSqlRawInvocation(invocation),
+            "ExecuteSqlRaw" or "ExecuteSqlRawAsync" => AnalyzeExecuteSqlRawInvocation(invocation),
+            "SqlQueryRaw" => AnalyzeSqlQueryRawInvocation(invocation),
+            _ => false
+        };
+
+        if (report)
+        {
+            context.ReportDiagnostic(Diagnostic.Create(
+                Descriptor,
+                GetTargetLocation(invocation.Syntax),
+                targetMethod.Name,
+                GetReplacementMethodName(targetMethod.Name)));
+        }
+
+        static Location GetTargetLocation(SyntaxNode syntax)
+        {
+            if (syntax is not InvocationExpressionSyntax invocationExpression)
+            {
+                Debug.Fail("In theory should never happen");
+                return syntax.GetLocation();
+            }
+
+            var targetNode = invocationExpression.Expression;
+
+            while (targetNode is MemberAccessExpressionSyntax memberAccess)
+            {
+                targetNode = memberAccess.Name;
+            }
+
+            // Generic name case, e.g. `db.Database.SqlQueryRaw<int>(...)`.
+            // At this point `targetNode` is `SqlQueryRaw<int>`, but we need location of the actual identifier
+            if (targetNode is GenericNameSyntax genericName)
+            {
+                return genericName.Identifier.GetLocation();
+            }
+
+            // We should appear at name expression, representing method name token, e.g.:
+            // db.Users.[|FromSqlRaw|](...) or db.Database.[|ExecuteSqlRaw|](...)
+            return targetNode.GetLocation();
+        }
+    }
+
+    internal static string GetReplacementMethodName(string oldName) => oldName switch
+    {
+        "FromSqlRaw" => "FromSql",
+        "ExecuteSqlRaw" => "ExecuteSql",
+        "ExecuteSqlRawAsync" => "ExecuteSqlAsync",
+        "SqlQueryRaw" => "SqlQuery",
+        _ => oldName
+    };
+
+    private static bool AnalyzeFromSqlRawInvocation(IInvocationOperation invocation)
+    {
+        var targetMethod = invocation.TargetMethod;
+        Debug.Assert(targetMethod.Name == "FromSqlRaw");
+
+        var compilation = invocation.SemanticModel!.Compilation;
+        var correctFromSqlRaw = compilation.FromSqlRawMethod();
+
+        if (correctFromSqlRaw is null)
+        {
+            Debug.Fail("Unable to find original `FromSqlRaw` method");
+            return false;
+        }
+
+        if (!targetMethod.ConstructedFrom.Equals(correctFromSqlRaw, SymbolEqualityComparer.Default))
+        {
+            return false;
+        }
+
+        // The second argument, that corresponds to `string sql` parameter, must be an interpolated string
+        if (invocation.Arguments[1].Value is not IInterpolatedStringOperation interpolatedString)
+        {
+            return false;
+        }
```

Comment:
Can do similar condensing into single if below as well.

```suggestion
        // The second argument, that corresponds to `string sql` parameter, must be an interpolated string
        if (correctFromSqlRaw is null
            || !targetMethod.ConstructedFrom.Equals(correctFromSqlRaw, SymbolEqualityComparer.Default)
            || invocation.Arguments[1].Value is not IInterpolatedStringOperation interpolatedString)
        {
            return false;
        }
```